### DISCUSSION TITLE
`AutoResetWrapper` does not, in fact, call `env.reset()`

### DISCUSSION THREAD

**Discussion by Theo-Cheynel** [google/brax#278] [General]
Hi,

The AutoResetWrapper (https://github.com/google/brax/blob/5ef4f654c63855dbc5f51543772feb348d48bb77/brax/envs/wrappers.py#L123) is supposedly a way to have the environment's reset function called when done is set to 1. However, it does not call the environment's `reset` method : instead, it uses the `qp` stored in `state.info['first_qp']` after the last external call to reset.

This assumes that the call to `env.reset` will always return the same value. However, in the example environments provided, there is some randomness introduced during the reset, which is not captured by the current AutoResetWrapper.

I think the expected behaviour is to have env.reset be called when needed, however I guess this might make things really slow. Could anything be done to fix this ?

Thanks

**Comment by btaba** [comment]
Hi @Theo-Cheynel ! The intention of AutoResetWrapper is indeed to use the `first_qp` and not to call `env.reset` for performance reasons as mentioned (since the whole thing needs to be jitted). This works in practice with a large batch size of environments

**Comment by Theo-Cheynel** [comment]
Thanks, that makes sense ! I think this has an impact on the training of my models though, since I have to set a moderate batch size in order to stay within my GPU's vRAM. Do you know what could be done to provide more randomness without impacting training performance too much ?

**Reply by btaba** [reply]
Hi @Theo-Cheynel ! You could try increasing the noise scale within your fixed batch size, or try running on multiple devices (the trainers run a pmap). Another idea is to call a real env.reset on the host periodically. Lastly, if the noise is simple enough to implement, you could also try modifying first_qp on the fly and see how that impacts performance. Let us know what you wind up doing!

**Reply by Theo-Cheynel** [reply]
Thanks ! For now I figured I'd simply call a true reset() every once in a while. In a near future I'll see if I can integrate the noise inside of the where_done, as it sounds like a better way to have a different starting qp everytime.
